### PR TITLE
More intuitive display of negative time deltas (#8274)

### PR DESCRIPTION
**Pull request by benvdh**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [x] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
This pull request changes the way negative time delta's are displayed. The default python display of a negative timedelta is rather unintuitive for a user of a dashboard. For example, -6 minutes is displayed as `'-1 day, 23:54:00`. This basically requires users to make time calculations in their heads to interpret the number correctly. This pull request changes the formatting to  '-0:06:00' at the code level and `-0 days 00:06:00` at the superset UI level to make it easier for users to interpret the time delta. More details on the issue and why this solution was chosen can be found in #8274 .

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
The data in postgres looks like this:
data_in_postgres

Before this pull request the data looks like this in superset:
data_in_superset_table_before

Afterwards it looks like this in superset:
data_in_superset_table_after

### TEST PLAN
- The unit test for base_json_conv() has been enhanced to handle timedelta's as well
- A new function has been added in superset.utils.core named `timedelta_f` which is called in base_json_conv() (same module). A new test has been added to test this formatting function.

### ADDITIONAL INFORMATION

- [x] Has associated issue: #8274 
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/utils/core.py
Comment (villebro) [inline superset/utils/core.py]: Please add type annotations

File: superset/utils/core.py
Comment (villebro) [inline superset/utils/core.py]: This recursion feels slightly unnecessary. Perhaps just doing `return "-"+ str(abs(td))` would be simpler?

**Review by villebro — CHANGES_REQUESTED** [review]
Left some comments. Also, do check out #8136 which might be relevant for you. I don't think it properly works in all charts yet, but if this is something that more people find useful I could probably work on improving support for durations.

File: superset/utils/core.py
Comment (benvdh) [reply superset/utils/core.py]: Good catch! Now fixed.

File: superset/utils/core.py
Comment (benvdh) [reply superset/utils/core.py]: That's indeed simpler, I thought the recursion did more than it actually did. Simplified now!

**Comment by benvdh** [comment]
> Left some comments. Also, do check out #8136 which might be relevant for you. I don't think it properly works in all charts yet, but if this is something that more people find useful I could probably work on improving support for durations.

Thanks for the quick review and helpful comments :+1: . They definitely improved the code. #8136 looks quite interesting and useful to my use case (described in the linked in the linked issue). Will play around with it when I find some time.

File: superset/utils/core.py
Comment (villebro) [inline superset/utils/core.py]: Perhaps `format_timedelta` might be more descriptive?

**Review by villebro — CHANGES_REQUESTED** [review]
Sorry for not catching this last pass, but having a more verbose name might be more clear. Other than that LGTM 👍

File: superset/utils/core.py
Comment (benvdh) [reply superset/utils/core.py]: @villebro I agree `format_timedelta` is clearer. It was actually what I used while developing,  until I noticed that there was already another function called `datetime_f` in the same module which did a similar thing for datetimes. Renaming would remove consistency in naming again. 

Would you still like me to make that change? (renaming both seems a bit beyond the scope of this pull request)

Comment (benvdh) [reply superset/utils/core.py]: Done anyway. We can always revert if necessary.

**Comment by benvdh** [comment]
@villebro I have renamed the function, but now suddenly the CI pipeline is failing on the cypress-sqllab test. This seems to be a known issue being addressed in open pull request #8215. Should we wait until this is fixed?

**Comment by villebro** [comment]
This is a known unrelated issue; I kicked off CI again.

**Comment by benvdh** [comment]
> 
> 
> This is a known unrelated issue; I kicked off CI again.

@villebro Thanks! Unfortunately, now it's the cypress-dashboard test (which passed in the previous built) with a similar unrelated timeOut issue:

```bash
 1) Dashboard top-level controls should allow chart level refresh:

     CypressError: Timed out retrying: cy.wait() timed out waiting 5000ms for the 1st request to the route: 'postJson_8_force'. No request ever occurred.
```
Are there anymore changes I need to make before this pull request can be approved?

**Review by villebro — APPROVED** [review]
LGTM. I'm not an expert on negative time durations, but this might be something that should be brought to the attention of the python core devs.

**Pull request merged**
